std/typecons.d

```diff
@@ -4865,12 +4865,154 @@ unittest
     assert(rc1 == 5);
     // No more allocation, add just one extra reference count
     auto rc2 = rc1;
+    assert(rc1.refCountedStore.refCount == 2);
     // Reference semantics
     rc2 = 42;
     assert(rc1 == 42);
     // the pair will be freed when rc1 and rc2 go out of scope
 }
 
+// recursively destroy a class instance
+private void _destroy(T)(T obj) if (is(T == class))
+{
+     foreach (C; TypeTuple!(T, BaseClassesTuple!T))
+     {
+         static if (__traits(hasMember, C, "__dtor"))
+             (cast(C)obj).__dtor();
+     }
+     // leak monitor b/c destruction would be impure, will be collected by the GC
+}
+
+/// RefCounted implementation for classes.
+struct RefCounted(T)
+    if (is(T == class))
+{
+    /// $(D RefCounted) storage implementation.
+    struct RefCountedStore
```

Comment:
Well, we overhauled `Unique` because it was plain bad, if it turns out we can find a better API here, we should do it.

I think we should stick as close to C++'s smart pointers as possible, but existing `Unique` and `RefCounted` both make that difficult. Their idiosyncratic behaviours make both smart pointers unsuitable for generic code and clumsy in any kind of code. This is mostly because of the inconsistent and/or poorly implemented null state, but I don't see a reason why `RefCounted` should expose the internal type used to put the value on the heap just to expose a couple of pieces of metadata. `isInitialized` wouldn't have existed in the first place if `RefCounted` was implemented as a proper smart pointer with a consistently expressed null state... the reference count is truly the only piece of metadata and thus the only thing that warrants a named member function in `RefCounted`.

I really don't know what the deal is with the whole auto-initialization thing. Lazily initialized objects tend to suck in practice, causing bugs where an intent to alias some object actually didn't because the reference was null at that time, despite there never being an explicit initialization anywhere to be found in the entire codebase. I've had issues with AAs behaving like this. Lazy initialization does *not* make reference types behave like value types.

/rant